pybossa/cache/__init__.py

```diff
@@ -212,6 +216,94 @@ def wrapper(*args, **kwargs):
     return decorator
 
 
+def memoize_with_l2_cache(timeout=DEFAULT_TIMEOUT,
+                          timeout_l2=L2_CACHE_TIMEOUT,
+                          timeout_mutex_lock=MUTEX_LOCK_TIMEOUT,
+                          cache_group_keys=None):
+    """
+    Decorator for caching functions using its arguments as part of the key.
+    Returns the cached value, or the function if the cache is disabled
+    If l1 cache miss, it will try to read l2 cache, which has a longer TTL
+    If l2 cache miss, it will try to obtain a mutex lock, read DB and
+    update l1 and l2 caches.
+    """
+    if timeout is None:
+        timeout = DEFAULT_TIMEOUT
+    if timeout < MIN_TIMEOUT:
+        timeout = MIN_TIMEOUT
+
```

Comment:
in order to avoid loop [here](https://github.com/bloomberg/pybossa/pull/696/files#diff-33727bfe074009fe81d9d050733000011157bfb907d47522884d3d975788111cR293) for longer time, I think we should limit to timeout_mutex_lock as
```python
if timeout_mutex_lock > MUTEX_LOCK_TIMEOUT:
      timeout_mutex_lock = MUTEX_LOCK_TIMEOUT 
```

Reply:
Thanks for the comment. The reason why we need a timeout for the mutex is to avoid the thread who is holding the mutex dies(e.g. server reboot) thus not releasing the mutex. So the timeout_mutex_lock is chosen by the client accordingly for the need. Only under the rare scenario that the loop you mentioned will loop to the end(normally, it should get the data from the cache [here](https://github.com/bloomberg/pybossa/pull/696/files#diff-33727bfe074009fe81d9d050733000011157bfb907d47522884d3d975788111cR296)). I feel it could be more flexible not to have the limit of ```timeout_mutex_lock``` to the ```MUTEX_LOCK_TIMEOUT``` since the client can decide what timeout_mutex_lock can be.